mesonbuild/interpreter/interpreter.py

```diff
@@ -2687,7 +2703,15 @@ def func_configure_file(self, node: mparser.BaseNode, args: T.List[TYPE_var],
                                      'copy a file to the build dir, use the \'copy:\' keyword '
                                      'argument added in 0.47.0', location=node)
             else:
-                mesonlib.dump_conf_header(ofile_abs, conf, output_format)
+                macro_name = kwargs['macro_name']
+                if not macro_name:
+                    if self.is_subproject():
+                        macro_path = os.path.join(os.path.relpath(ofile_path, 'subprojects'), ofile_fname)
```

Comment:
On the heuristics topic...

The subproject directory accepts a project() argument for overriding it, e.g. if existing codebases already have a third_party directory that they want to keep using.